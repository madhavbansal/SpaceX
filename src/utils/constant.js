export const constants = {

  spaceXPrograms:  'SpaceX Launch Programs',
  filters:'Filters',
  missionIds:'Mission Ids:',
  launchYear:'Launch Year:',
  successfulLaunch:'Successful Launch:',
  successfulLanding:'Successful Landing:'



}