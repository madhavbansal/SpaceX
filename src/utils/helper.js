export const sectionConfig = {
    launchYear: {
        sectionName: "Launch Year",
        buttons: [{
            id: '2006',
            label: '2006'
        }, {
            id: '2007',
            label: '2007'
        }, {
            id: '2008',
            label: '2008'
        }, {
            id: '2009',
            label: '2009'
        }, {
            id: '2010',
            label: '2010'
        }, {
            id: '2011',
            label: '2011'
        }, {
            id: '2012',
            label: '2012'
        }, {
            id: '2013',
            label: '2013'
        }, {
            id: '2014',
            label: '2014'
        }, {
            id: '2015',
            label: '2015'
        }, {
            id: '2016',
            label: '2016'
        }, {
            id: '2017',
            label: '2017'
        }, {
            id: '2018',
            label: '2018'
        }, {
            id: '2019',
            label: '2019'
        }, {
            id: '2020',
            label: '2020'
        }],
        filterName: 'launchYear'
    },
    sucessfulLaunch: {
        sectionName: "Successful Launch",
        buttons: [{ id: 'true', label: 'True' }, { id: 'false', label: 'False' }],
        filterName: 'sucessfulLaunch'
    },
    sucessfulLanding: {
        sectionName: "Successful Landing",
        buttons: [{ id: 'true', label: 'True' }, { id: 'false', label: 'False' }],
        filterName: 'sucessfulLanding'
    }

};

// used to construct URL in FilterSection.js
export const constructUrl = (baseUrl, params) => {
    const keys = Object.keys(params);
    return keys.reduce((prevValue, currentValue, idx) => {
        let str = idx ? `&${currentValue}=${params[currentValue]}` : `${currentValue}=${params[currentValue]}`;
        return `${prevValue}${str}`

    }, `${baseUrl}?`);

}

export const filterMapping = {
    launchYear: 'launch_year',
    sucessfulLanding: 'land_success',
    sucessfulLaunch: 'launch_success'

};

export const constructApiUrl = (baseUrl, params) => {
    const keys = Object.keys(params);
    return keys.reduce((prevValue, currentValue, idx) => {
        let str = `&${filterMapping[currentValue]}=${params[currentValue]}`;
        return `${prevValue}${str}`

    }, `${baseUrl}`);

}

export const baseUrl = 'https://api.spacexdata.com/v3/launches?limit=100';


