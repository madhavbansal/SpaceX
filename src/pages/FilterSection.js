import React, { useState } from 'react';
import { useLocation, useHistory } from "react-router-dom";
import { ContainedButtons } from '../components';
import queryString from 'query-string';
import { constructUrl } from '../utils/helper';

const FilterSection = (props) => {
    const history = useHistory();
    const location = useLocation();
    const [selectedButtonConfig, setSelectedButtonConfig] = useState({});
    const buttonCLickHandler = (id, filterName) => {
        let params = queryString.parse(location.search, { parseBooleans: true });
        params = { ...params, [filterName]: id };
        const newUrl = constructUrl('/', params);
        setSelectedButtonConfig({...params});
        history.push(newUrl);
    }

    const { sectionConfig: { sectionName, buttons, filterName } } = props;

    return (
        <>

            {sectionName}
            {buttons.map(({ label, id }) => {
            const isSelected = selectedButtonConfig[filterName]===id;
            return(
                <ContainedButtons label={label} id={id} key={id} isSelected={isSelected} onClick={() => buttonCLickHandler(id, filterName)} />
            )})}
        </>
    )


}

export default FilterSection;