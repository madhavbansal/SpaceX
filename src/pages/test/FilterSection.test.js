import React from 'react';
import renderer from 'react-test-renderer';
import FilterSection from '../FilterSection';
import { sectionConfig } from '../../utils/helper';


jest.mock('../../components/Button', () => 'ContainedButtons');
jest.mock('react-router-dom', () =>{
return {
    useLocation:()=>({}),
    useHistory:()=>({})
}
});

test('Test the FilterSection Component', () => {
    const component = renderer.create(
        <FilterSection sectionConfig={sectionConfig.launchYear} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});