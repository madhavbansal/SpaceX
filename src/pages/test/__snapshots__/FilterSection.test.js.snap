// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the FilterSection Component 1`] = `
Array [
  "Launch Year",
  <ContainedButtons
    id="2006"
    isSelected={false}
    label="2006"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2007"
    isSelected={false}
    label="2007"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2008"
    isSelected={false}
    label="2008"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2009"
    isSelected={false}
    label="2009"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2010"
    isSelected={false}
    label="2010"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2011"
    isSelected={false}
    label="2011"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2012"
    isSelected={false}
    label="2012"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2013"
    isSelected={false}
    label="2013"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2014"
    isSelected={false}
    label="2014"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2015"
    isSelected={false}
    label="2015"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2016"
    isSelected={false}
    label="2016"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2017"
    isSelected={false}
    label="2017"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2018"
    isSelected={false}
    label="2018"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2019"
    isSelected={false}
    label="2019"
    onClick={[Function]}
  />,
  <ContainedButtons
    id="2020"
    isSelected={false}
    label="2020"
    onClick={[Function]}
  />,
]
`;
